## Data Understanding

##### Column decsriptions


1. Contract Number: 66,549 unique values (slightly fewer than the total rows), with 72 missing entries.
2. Amount: 50,594 unique values, indicating a high variance in contract amounts, with 855 missing values.
3. Dates:
Various date columns have a broad range of unique values, suggesting contracts are spread over a significant timeline.
4. Sign Date, Start Date, End Date, etc., have varying numbers of unique values, hinting at different stages and lengths of contract processes.
5. Agpo Certificate Number: Only 13,606 unique values, with 57,311 missing entries, indicating that this information may not be recorded for every contract.
6. Awarded Agpo Group Id: Contains only 4 unique values (Women, Youth, All, People with Disabilities) but has a high number of missing values (58,261).
7. Terminated: Only one unique value, 1.0, across the 88 non-null entries, suggesting limited use of this column in identifying terminated contracts.
8. Financial Year: 15 unique values, representing different fiscal years.
9. Quarter: 4 unique values (Q1, Q2, Q3, Q4), with 3,495 missing entries.
10. PE Name: 498 unique values, indicating a range of entities participating in the contracts.
11. Supplier Name: 25,419 unique supplier names, suggesting substantial diversity in supplier entities.
12. No. of B.O.I: Contains only 19 unique values, with the majority of entries missing (67,866).
13. Created At: Timestamped information on record creation with 48,558 unique values and 34,028 missing values, hinting at possibly incomplete time-tracking data.

In [1]:
import pandas as pd
import numpy as np
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import LabelEncoder
from datetime import datetime
import matplotlib.pyplot as plt

# Load the dataset
df = pd.read_excel("published_contracts (3).xlsx")

In [2]:
df.head()

Contract Number     Amount  Award Date  \
0          0015.1  1962488.0  2018-09-26   
1           38632   257736.0  2018-10-02   
2           38631   254736.0  2018-09-25   
3            0189   149300.0  2019-02-19   
4       LPO 1079.    81500.0  2018-12-27   

                                        Tender Title Eval Completion Date  \
0  Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-09-25   
1  Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-09-25   
2  Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-09-19   
3  Supply of Non Pharmaceuticals (Surgical Tubes,...           2019-01-15   
4  Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-12-27   

  Notification Of Award Date   Sign Date  Start Date    End Date  \
0                 2018-09-26  2018-09-26  2018-09-26  2018-10-25   
1                 2018-10-04  2018-10-05  2018-10-08  2018-10-23   
2                 2018-10-02  2018-10-05  2018-10-12  2018-11-12   
3                 2019-02-07  2019-02-11  2019-02-11  2019-03-11   
4                 2018-12-27  2018-12-27  2018-12-27  2019-01-26   

  Agpo Certificate Number Awarded Agpo Group Id  Created By  Terminated  \
0                     NaN                   NaN           1         NaN   
1                     NaN                   NaN           1         NaN   
2                     NaN                   NaN           1         NaN   
3                     NaN                   NaN           1         NaN   
4                     NaN                   NaN           1         NaN   

  Financial Year Quarter             Tender Ref.  \
0      2018/2019      Q1  KEMSA/ONT 06/2017-2019   
1      2018/2019      Q2  KEMSA/ONT 06/2017-2019   
2      2018/2019      Q1  KEMSA/ONT 06/2017-2019   
3      2018/2019      Q3  KEMSA/ONT 06/2017-2019   
4      2018/2019      Q2  KEMSA/ONT 06/2017-2019   

                         PE Name                       Supplier Name  \
0    Higher Education Loan Board       VIABLE DECO SOLUTIONS LIMITED   
1  Kenya Post Office Saving Bank                COMPUTERWAYS LIMITED   
2  Kenya Post Office Saving Bank  REALISTIC SYSTEMS AND TECHNOLOGIES   
3                           Embu    JIMRIVER CARGO LOGISTICS LIMITED   
4    Higher Education Loan Board                   MATKY INVESTMENTS   

   No. of B.O.I           Created At  
0           NaN  2019-01-31 09:55:20  
1           NaN  2019-02-19 10:14:13  
2           NaN  2019-02-27 14:28:26  
3           NaN  2019-03-15 10:03:54  
4           NaN  2019-03-17 18:11:54

In [3]:
# Shape of the dataset
df.shape

(82708, 20)

In [4]:
# Checking for null values
df.isna().sum()

Contract Number                  72
Amount                          855
Award Date                        0
Tender Title                     15
Eval Completion Date              0
Notification Of Award Date        0
Sign Date                     12724
Start Date                      988
End Date                       1012
Agpo Certificate Number       57311
Awarded Agpo Group Id         58261
Created By                        0
Terminated                    82620
Financial Year                    0
Quarter                        3495
Tender Ref.                      15
PE Name                           0
Supplier Name                     0
No. of B.O.I                  67866
Created At                    34028
dtype: int64

In [5]:
# Checkng for duplicates
df.duplicated().sum()

141

In [6]:
import pandas as pd # Importing pandas for data manipulation

class DataUnderstanding():
    """Class that provides an understanding of a dataset"""

    def __init__(self, data=None):
        """Initialization"""
        self.df = data

    def load_data(self, path):
        """Load the data"""
        if self.df is None:

            self.df = pd.read_excel(path)
        return self.df

    def concat_data(self, other_df):
        """Concatenate the current dataframe with another dataframe vertically"""
        if self.df is not None and other_df is not None:
            self.df = pd.concat([self.df, other_df], axis=0, ignore_index=True)
        return self.df

    def understanding(self):
        """Provides insights into the dataset"""
        # Info
        print("INFO")
        print("-" * 4)
        self.df.info()

        # Shape
        print("\n\nSHAPE")
        print("-" * 5)
        print(f"Records in dataset: {self.df.shape[0]} with {self.df.shape[1]} columns.")

        # Columns
        print("\n\nCOLUMNS")
        print("-" * 6)
        print("Columns in the dataset are:")
        for idx in self.df.columns:
            print(f"- {idx}")

        # Unique Values
        print("\n\nUNIQUE VALUES")
        print("-" * 12)
        for col in self.df.columns:
            print(f"Column {col} has {self.df[col].nunique()} unique values")
            if self.df[col].nunique() < 12:
                print(f"Top unique values in {col} include:")
                for idx in self.df[col].value_counts().index:
                    print(f"- {idx}")
            print("")

        # Missing or Null Values
        print("\nMISSING VALUES")
        print("-" * 15)
        for col in self.df.columns:
            print(f"Column {col} has {self.df[col].isnull().sum()} missing values.")

        # Duplicate Values
        print("\n\nDUPLICATE VALUES")
        print("-" * 16)
        print(f"The dataset has {self.df.duplicated().sum()} duplicated records.")

# Initialize data understanding
data = DataUnderstanding()

# Load the first dataset
data_path1 = "published_contracts (3).xlsx" # The path to your Excel file
df = data.load_data(data_path1)


df

Contract Number     Amount  Award Date  \
0              0015.1  1962488.0  2018-09-26   
1               38632   257736.0  2018-10-02   
2               38631   254736.0  2018-09-25   
3                0189   149300.0  2019-02-19   
4           LPO 1079.    81500.0  2018-12-27   
...               ...        ...         ...   
82703         4298079        NaN  2024-07-25   
82704         4298095    79700.0  2024-07-24   
82705         4298092    69800.0  2024-07-25   
82706         1761335    55700.0  2024-07-25   
82707         1761331    47100.0  2024-07-25   

                                            Tender Title Eval Completion Date  \
0      Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-09-25   
1      Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-09-25   
2      Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-09-19   
3      Supply of Non Pharmaceuticals (Surgical Tubes,...           2019-01-15   
4      Supply of Non Pharmaceuticals (Surgical Tubes,...           2018-12-27   
...                                                  ...                  ...   
82703                                     SUPPLY OF FUEL           2024-07-24   
82704                       SUPPLY OF CATERING MATERIALS           2024-07-24   
82705                               SUPPLY OF STARIONERY           2024-07-24   
82706                                 REPAIR OF KDK 171P           2024-07-21   
82707                                 REPAIR OF KDA 671P           2024-07-24   

      Notification Of Award Date   Sign Date  Start Date    End Date  \
0                     2018-09-26  2018-09-26  2018-09-26  2018-10-25   
1                     2018-10-04  2018-10-05  2018-10-08  2018-10-23   
2                     2018-10-02  2018-10-05  2018-10-12  2018-11-12   
3                     2019-02-07  2019-02-11  2019-02-11  2019-03-11   
4                     2018-12-27  2018-12-27  2018-12-27  2019-01-26   
...                          ...         ...         ...         ...   
82703                 2024-07-26  2024-09-06  2024-09-06  2024-09-06   
82704                 2029-07-24  2024-09-19  2024-09-20  2024-09-20   
82705                 2024-07-25  2024-07-26  2024-09-10  2024-09-10   
82706                 2024-07-25  2024-09-23  2024-09-23  2024-09-27   
82707                 2024-07-26  2024-09-12  2024-09-13  2024-09-13   

      Agpo Certificate Number Awarded Agpo Group Id  Created By  Terminated  \
0                         NaN                   NaN           1         NaN   
1                         NaN                   NaN           1         NaN   
2                         NaN                   NaN           1         NaN   
3                         NaN                   NaN           1         NaN   
4                         NaN                   NaN           1         NaN   
...                       ...                   ...         ...         ...   
82703                     NaN                   NaN       15737         NaN   
82704                     NaN                   NaN       15737         NaN   
82705                     NaN                   NaN       15737         NaN   
82706                     NaN                   NaN       15737         NaN   
82707                     NaN                   NaN       15737         NaN   

      Financial Year Quarter             Tender Ref.  \
0          2018/2019      Q1  KEMSA/ONT 06/2017-2019   
1          2018/2019      Q2  KEMSA/ONT 06/2017-2019   
2          2018/2019      Q1  KEMSA/ONT 06/2017-2019   
3          2018/2019      Q3  KEMSA/ONT 06/2017-2019   
4          2018/2019      Q2  KEMSA/ONT 06/2017-2019   
...              ...     ...                     ...   
82703      2024/2025      Q1     NDMA/SBU/OO1A/24-24   
82704      2024/2025      Q1     NDMA/SBU/004A/24-25   
82705      2024/2025      Q1     NDMA/SBU/OO3A/24-25   
82706      2024/2025      Q1  NDMA/SBU/015/2024-2025   
82707      2024/2025      

In [7]:
# Get an understanding of the dataset
data.understanding()

INFO
----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82708 entries, 0 to 82707
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Contract Number             82636 non-null  object 
 1   Amount                      81853 non-null  float64
 2   Award Date                  82708 non-null  object 
 3   Tender Title                82693 non-null  object 
 4   Eval Completion Date        82708 non-null  object 
 5   Notification Of Award Date  82708 non-null  object 
 6   Sign Date                   69984 non-null  object 
 7   Start Date                  81720 non-null  object 
 8   End Date                    81696 non-null  object 
 9   Agpo Certificate Number     25397 non-null  object 
 10  Awarded Agpo Group Id       24447 non-null  object 
 11  Created By                  82708 non-null  int64  
 12  Terminated                  88 non-null     float64
 13  Financial Year       

#### Summary Observations:
- Data Completeness: Several columns, especially those related to specific certifications and awards (e.g., Agpo Certificate Number and Awarded Agpo Group Id), have significant missing values, which may affect analyses focused on these categories.
- Data Quality: With 141 duplicates and varying levels of missing values, some data cleaning will be required, particularly for categorical columns that are missing a large portion of data.
- Potential Areas for Analysis: Despite missing values, substantial information is available regarding contract details (e.g., Amount, PE Name, Supplier Name), dates, and contract progression stages. These can support analyses related to contract timelines, supplier diversity, and financial tracking over different periods (fiscal years and quarters).

The dataset has 124 duplicate values


Here's an analysis of each column's relevance for fraud detection, ranked by importance:
High Relevance (Critical for Fraud Detection):

Temporal Columns (Pattern Analysis):


Award Date
Eval Completion Date
Notification Of Award Date
Sign Date
Start Date
End Date
→ These can reveal suspicious patterns like:

Backdating
Unrealistic timelines
Process sequence violations




Financial Information:


Amount
→ Critical for detecting:

Split purchases
Threshold avoidance
Unusual pricing patterns
Round numbers (often indicative of fraud)




Entity Information:


Supplier Name
PE Name (Procuring Entity)
→ Important for:

Shell company detection
Multiple awards to same supplier
Cross-referencing with known fraudulent entities



Medium Relevance:

Process Identifiers:


Contract Number
Tender Ref.
→ Useful for:

Detecting duplicate contracts
Sequential pattern analysis
Reference number manipulation




Classification Fields:


Financial Year
Quarter
→ Helpful for:

End-of-period fraud patterns
Budget exhaustion patterns
Seasonal anomalies




Compliance Indicators:


Agpo Certificate Number
Awarded Agpo Group Id
→ Can indicate:

Certificate fraud
Qualification manipulation



Lower Relevance (But Still Useful):

Metadata:


Created By
Created At
No. of B.O.I
→ Can reveal:

Unusual creation patterns
User behavior anomalies




Status Indicators:


Terminated
→ May indicate:

Contract execution issues
Post-award problems



For your ML model, I recommend focusing on these key feature engineering approaches:

Temporal Features:


Time differences between dates
Process duration
Date sequence violations
End of period patterns


Financial Features:


Amount distributions
Threshold proximity
Round number detection
Split payment patterns


Entity Features:


Supplier award frequency
Entity relationships
New supplier flags
Geographic patterns


Metadata Features:


Creation patterns
User behavior patterns
Documentation completeness

## Data Cleaning

In [ ]:
class DataCleaning(DataUnderstanding):
    """This class is used for data cleaning"""

    def drop_columns(self, columns):
        """Drop specified columns"""
        if self.df is not None:
            print(f"Dropping Columns: {columns}")
            self.df.drop(columns=columns, inplace=True)


    def strip_column_names(self):
        """Strip whitespace from column names"""
        if self.df is not None:
            print("Stripping whitespace from column names")
            self.df.columns = self.df.columns.str.strip()

    def convert_to_datetime(self, column_name):
        """Convert a column to datetime format"""
        if self.df is not None:
            self.df[column_name] = pd.to_datetime(self.df[column_name])

    def filter_year(self, column_name, year):
        """Filter rows based on a specific year"""
        if self.df is not None:
            self.df = self.df[self.df[column_name].dt.year == year]

    def num_duplicates(self,df):
        """Check for the total number of duplicates"""
        if self.df is not None:
            print(f"Number of duplicates in {self.df.duplicated().sum()}")
            self.df.duplicated().sum()


data = DataCleaning()

# Load the first dataset
data_path1 = 'OlympicsData2.csv'
data.load_data(data_path1)

# Load the second dataset
data_path2 = 'OlympicsData1.csv'
df2 = pd.read_csv(data_path2, encoding='latin-1')

# Concatenate the two datasets vertically
data.concat_data(df2)

# Perform data cleaning
# Drop irrelevant columns
irrelevant_columns = ['Tweet_Image_URL', 'Web_Page_URL', 'Author_Web_Page_URL', 'Author_Name',
                      'Tweet_Video_URL', 'Tweet_AD', 'Tweet_Website','Tweet_Number_of_Reviews']
data.drop_columns(irrelevant_columns)

# Strip column names of any leading/trailing whitespace
data.strip_column_names()

# Convert 'Tweet_Timestamp' to datetime format
data.convert_to_datetime('Tweet_Timestamp')

# Store the final cleaned dataset in df
df = data.df

# Print the cleaned DataFrame
print(df.head())